### PR TITLE
Calendly: Updated the rendering of the styles to use the 'style' attribute

### PR DESCRIPTION
**Pull request by pablinos**
When rendering the block colours on the server, we were using an
`inline_style` that rendered as a style tag. This meant that each style
tag would override the last, and we were also using the wrong selector,
so the colours weren't being applied on the front end.

#### Changes proposed in this Pull Request:
This change adds the styles to each button as an inline style attribute,
so they're unique to each instance.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
This is a fix to a new feature

#### Testing instructions:
* Create a post with two calendly blocks
* Set them to the button style
* Make them have different colours
* Publish/Preview the post and check that the buttons are displayed with the correct colours

#### Proposed changelog entry for your changes:
Not required

### CODE REVIEW

**Comment by glendaviesnz** [comment]
Tests ok for me, except the there is an issue with the text colour on hover:

hover

If dark background is selected, but no text colour, then the above happens, but if you select a background and a text colour the text colour doesn't change on hover.

**Comment by pablinos** [comment]
Ah, yes I think I've seen that problem previously. I'm wondering if it wasn't showing up when the colours were applied via a rule in a style tag, as they inherited correctly. 

I'll have a ponder too!

**Comment by pablinos** [comment]
I've reverted to adding a style tag and used an ID to target the block. It works in my tests. Hopefully, it will for you too @glendaviesnz !

**Comment by glendaviesnz** [comment]
This only seems to work for me if the text colour as well as background colour is set. If only a background colour is selected the hover issues is still there:

Only background specified:

no-text-colour

Text and background specified:

text-colour

This is on TwentyTwenty, which specifies the text hover should just fall back to the body colour. We could add a :hover to the id targeted inline style if no text colour is set, but I can't see an easy way of working out what that hover colour should be :-(

**Comment by pablinos** [comment]
Thanks for looking at it again @glendaviesnz. It's pretty annoying, isn't it? I'm beginning to think this is more of a theme issue, but one option could be to create our own `wp-block-button__link` class that duplicates just the styles we need.

With this and the alignment stuff, I think we may need to have tests for themes that make sure blocks can be aligned properly and colours can be set. Having links go back to the body colour once they've been visited seems like something we should discuss changing.

File: extensions/blocks/calendly/calendly.php [hunk 1]
```diff
@@ -138,8 +140,9 @@ function load_assets( $attr, $content ) {
 		}
 
 		$content = sprintf(
-			'<div class="%1$s"><a class="wp-block-button__link" role="button" onclick="Calendly.initPopupWidget({url:\'%2$s\'});return false;">%3$s</a></div>',
+			'<div class="%1$s" id="%2$s"><a class="wp-block-button__link" role="button" onclick="Calendly.initPopupWidget({url:\'%3$s\'});return false;">%4$s</a></div>',
 			esc_attr( $classes ),
+			$block_id,
```
Comment (jeherve) [inline extensions/blocks/calendly/calendly.php]: Can we escape this?

```suggestion
			esc_attr( $block_id ),
```

File: extensions/blocks/calendly/calendly.php [hunk 2]
```diff
@@ -126,7 +127,8 @@ function load_assets( $attr, $content ) {
 		 */
 		if ( ! empty( $submit_button_text_color ) || ! empty( $submit_button_background_color ) ) {
 			$inline_styles = sprintf(
-				'.wp-block-jetpack-calendly .button{%1$s%2$s}',
+				'#%1$s .wp-block-button__link{%2$s%3$s}',
+				$block_id,
```
Comment (jeherve) [inline extensions/blocks/calendly/calendly.php]:
```suggestion
				esc_attr( $block_id ),
```

**Review by jeherve — CHANGES_REQUESTED** [review]

**Review by jeherve — APPROVED** [review]

**Comment by Copons** [comment]
Is there a reason why we are using `wp_add_inline_style` instead of _literally_ adding inline style to the rendered HTML?

We used to to that in FSE, inspired by some Core blocks, and then Core did the same for the Navigation block (I guess we cross-inspired each other 😄):

https://github.com/Automattic/wp-calypso/blob/c3480ac6ab84378f7c528e841c1647dfe4277264/apps/full-site-editing/full-site-editing-plugin/dotcom-fse/blocks/site-title/index.php

https://github.com/WordPress/gutenberg/blob/7dda20a4610102df982a60a3c5f8b1569e0576f6/packages/block-library/src/navigation/index.php

tl;dr it would be something like this:

```php
function render_callback( $attrs ) {
  $class = 'wp-block-foobar-class'
  $style = '';

  if ( isset( $attributes['textColor'] ) {
    $class .= ' has-text-color';
    $class .= ' has-' . $attributes['textColor'] . '-color';
  } elseif ( isset( $attributes['customTextColor'] ) ) {
    $class .= ' has-text-color';
    $style .= ' color: ' . $attributes['customTextColor'] . ';';
  }

return '<div class="' . esc_attr( $class ) . '" style="' . esc_attr( $style ) . '">The Block</div>';
}
```

**Comment by jeherve** [comment]
> Is there a reason why we are using wp_add_inline_style instead of literally adding inline style to the rendered HTML?

I like that we use `wp_add_inline_style`, it makes it easier for folks who may want to overwrite or remove those styles in the future. This may be something we need to do on our end as well, when building a custom solution for AMP for example. I think it's a more responsible approach in plugins and themes; inline styles aren't easy to clean up or overwrite in general.

**Comment by jeherve** [comment]
Cherry-picked to `branch-8.2` in f3381f6543baa02f12e94749dab3bd2d74e92bc5

**Comment by Copons** [comment]
> > Is there a reason why we are using wp_add_inline_style instead of literally adding inline style to the rendered HTML?
> 
> I like that we use `wp_add_inline_style`, it makes it easier for folks who may want to overwrite or remove those styles in the future. This may be something we need to do on our end as well, when building a custom solution for AMP for example. I think it's a more responsible approach in plugins and themes; inline styles aren't easy to clean up or overwrite in general.

I'm not sure how much easier it is to override a rule targeting a randomized ID though 😄 

The approach I've proposed instead is very overridable.

When the user uses the theme's predefined colors, we add standard Gutenberg classes (not inline styles!) to the element. E.g. `has-text-color has-primary-color` or `has-background-color has-primary-background-color`
The burden of maintaining these classes is on the theme, and it's a good thing! If a [theme declares support for a color palette](https://developer.wordpress.org/block-editor/developers/themes/theme-support/#block-color-palettes), they are also supposed to have those CSS classes defined anyway.

To be more clear:

- With the current `wp_add_inline_style` approach:
```css
#calendly-block-123 .wp-block-button__link {
  color: #ffffff;
  background-color: #000000;
}
```

- With the Gutenberg classes approach:
```html
<a
  class="wp-block-button__link has-text-color has-light-color has-background-color has-dark-background-color"
>Link</a>
```

On the other hand, if the user picks a _custom_ color, we shouldn't use a class (custom color is not supposed to be reusable), and therefore fall back to the inline style.
```html
<a
  class="wp-block-button__link has-text-color has-background-color"
  style="color: #123456; background-color: #ABCDEF"
>Link</a>
```

**Comment by jeherve** [comment]
> I'm not sure how much easier it is to override a rule targeting a randomized ID though

I think in those cases, one wanting to override things would just remove the whole set of styles added by Jetpack and add their own instead.

> When the user uses the theme's predefined colors, we add standard Gutenberg classes (not inline styles!) to the element. E.g. has-text-color has-primary-color or has-background-color has-primary-background-color

Yes, I think that's the best way, and as you mentioned, the most theme-friendly. I think we should aim for that whenever possible.

However, when adding custom colors, and since we can't rely on that approach, I would still vote for `wp_add_inline_style` instead of inline styles.

**Comment by Copons** [comment]
@jeherve I'm so sorry to be so annoying during release day! 😅 

> However, when adding custom colors, and since we can't rely on that approach, I would still vote for `wp_add_inline_style` instead of inline styles.

I don't really see the value of letting the user override attributes they could fully customize by themselves.
I understand that inline styles are not cool, but in this particular case (custom styles that intrinsically can't be "_class_-ified") they are totally acceptable.
Also, imho they are much more transparent than assigning a random ID to all these blocks, and createing an additional `<style>` tag that targets that random ID.

**Comment by jeherve** [comment]
@Copons I am not married to the idea; it feels more "WordPress-y" to me to use `wp_add_inline_style`, as it plays well with others and is more compatible with things like caching / concatenating plugins, but we can definitely use inline styles instead. Either way, it won't stop anyone from customizing things with an `!important`.

**Comment by jeherve** [comment]
r202925-wpcom
